### PR TITLE
NFSP CNN + Updated Imperfect Recall Dark Hex

### PR DESCRIPTION
**Pull request by BedirT**
NFSP part is not fully complete, mainly uploading not to have multiple PRs for the same purpose.

IR-DarkHex updates work fine on the tests. When trying to generate new playthroughs, I get a weird shape on `ObservationTensorShape`,  It works well when I remove the function and inherit the old one (even though the numbers are wrong in that case), so I think the issue is on

https://github.com/deepmind/open_spiel/blob/7e6473cc6e64b1131b79553e4823ff3910f58c95/open_spiel/games/dark_hex.cc#L341-L343

I am not using `num_cells_` since it acts weird when I make it public or protected instead of private. Addressing #789

### CODE REVIEW

File: open_spiel/games/dark_hex.cc [hunk 1]
```diff
@@ -322,6 +322,42 @@ std::vector<int> DarkHexGame::ObservationTensorShape() const {
...
+
+  const auto& player_view = (player == 0 ? black_view_ : white_view_);
+  for (int cell = 0; cell < num_cells(); ++cell) {
+    values[cell] = static_cast<int>(player_view[cell]) - kMinValueCellState;
+  }
+}
+
+std::vector<int> ImperfectRecallDarkHexGame::ObservationTensorShape() const {
+  return {num_cells()};
+}
```
Comment (lanctot) [inline open_spiel/games/dark_hex.cc:360]: These should be multi-dimensional. 

See Breakthrough and Tic-Tac-Toe. Use the TensorView helper to make planar observations as done in those games.

File: open_spiel/games/dark_hex.cc [hunk 2]
```diff
@@ -323,5 +325,63 @@ std::vector<int> DarkHexGame::ObservationTensorShape() const {
...
+      SPIEL_CHECK_TRUE(plane >= 0 && plane < 3);
+      view[{plane, r, c}] = 1.0;
+    }
+  }
+}
+
+void ImperfectRecallDarkHexState::ObservationTensor(Player player,
+                                     absl::Span<float> values) const {
+  // Same as the InformationStateTensor.
+  InformationStateTensor(player, values);
```
Comment (BedirT) [inline open_spiel/games/dark_hex.cc:379]: Not sure if this is legal for OS, but it is unnecessary to repeat the code I think.

**Comment by BedirT** [comment]
Yeah, I am using the updated ones for my experiments right now, but couldn't find time to clean up. Will do it in couple of weeks hopefully.

Yeah subclass will be better I think, I will do that.

**Comment by lanctot** [comment]
Closing for now, feel free to re-open a new PR.

**Pull request closed without merging**